cli/formatter/go-pretty/extra/001_add_transposing_and_colconf.patch

```diff
@@ -0,0 +1,53 @@
+diff --git a/table/table.go b/table/table.go
+index 684974c..55dafb9 100644
+--- a/table/table.go
++++ b/table/table.go
+@@ -158,6 +158,36 @@ func (t *Table) AppendRows(rows []Row, config ...RowConfig) {
+ 	}
+ }
+ 
++// TransposeAppendedRows transposes the appended rows.
++func (t *Table) TransposeAppendedRows() {
++	var rowsRawTransposedCap = 0
++	for _, row := range t.rowsRaw {
++		if len(row) > rowsRawTransposedCap {
++			rowsRawTransposedCap = len(row)
++		}
++	}
++
++	for rowNum, row := range t.rowsRaw {
++		if len(row) < rowsRawTransposedCap {
++			for i := 0; i < rowsRawTransposedCap-len(row); i++ {
++				t.rowsRaw[rowNum] = append(t.rowsRaw[rowNum], "")
++			}
++		}
++	}
++
++	var rowsRawTransposed []Row
++
++	for i := 0; i < rowsRawTransposedCap; i++ {
++		var rowTransposed Row
++		for j := 0; j < len(t.rowsRaw); j++ {
++			rowTransposed = append(rowTransposed, t.rowsRaw[j][i])
++		}
++		rowsRawTransposed = append(rowsRawTransposed, rowTransposed)
++	}
++
++	t.rowsRaw = rowsRawTransposed
++}
```

Comment:
Can we transpose lines in the client code without patching the library before appending it into a table? Why can't this be done? I don't see the reason at first glance.

Reply:
Unfortunately not, this method works with private entities (`t.rowsRaw`)

Reply:
If we have already appended rows. But what about transpose before appending? The idea in pseudo-code: 

```Go
var []rows table.Row

// Build rows slice code.
// ...
// Done.

if opts.TransposeMode {
    rows := someTransposeRowsMethod(rows)
}

t := table.NewWriter()
for _, r := range rows {
    t.AppendRow(r)
}
```
or
```Go
var []srcRows someType

// Build srcRows slice code.
// ...
// Done.

if opts.TransposeMode {
    srcRows := someTransposeRowsMethod(srcRows)
}

t := table.NewWriter()
for _, r := range rows {
    var row table.Row
    row := srcToRow(r)   
    t.AppendRow(row)
}
```

Reply:
I thought about it, unfortunately it is impossible: the logic of vertical tables merging and the design of filling the table with rows does not allow this operation to be performed earlier

We can try to upstream this method later

Reply:
> the logic of vertical tables merging and the design of filling the table with rows does not allow this operation to be performed earlier

Could you please give an example with the problem?

Reply:
When the table is transposed, the rows become columns. We need to have a complete set of rows to perform the transposition. And the addition to the table is performed not by columns, but by rows (otherwise it would really be simple)

Reply:
What wrong with the patch?
```patch
--- a/cli/formatter/format.go
+++ b/cli/formatter/format.go
@@ -214,7 +214,9 @@ func renderMapsWithEqualKeys(maps []map[string]renderNode, opts *Opts) (string,
     sort.Strings(commonKeys)
 
     t := table.NewWriter()
-    t.AppendRow(constructHeaderRowForMap(commonKeys))
+    var rows []table.Row
+
+    rows = append(rows, constructHeaderRowForMap(commonKeys))
 
     for _, mapVal := range maps {
         var rowVals table.Row
@@ -238,14 +240,16 @@ func renderMapsWithEqualKeys(maps []map[string]renderNode, opts *Opts) (string,
                 rowVals = append(rowVals, string(jsonVal))
             }
         }
-        t.AppendRow(rowVals)
+        rows = append(rows, rowVals)
     }
 
     t.Style().Options.SeparateRows = true
-    var rowsAmount = t.Length()
+    var rowsAmount = len(rows)
     if opts.TransposeMode {
-        t.TransposeAppendedRows()
+        rows = transposeRows(rows)
     }
+    t.AppendRows(rows)
+
     if opts.NoGraphics {
         t.SetStyle(table.Style{Box: StyleWithoutGraphics})
     }
@@ -363,8 +367,9 @@ func constructMarkdownEmptyRow(totalRowAmount int) string {
 // renderSingleColumnTable returns table with single column as string.
 func renderSingleColumnTable(renderBatch []renderNode, opts *Opts) string {
     t := table.NewWriter()
+
     var rows []table.Row
-    t.AppendRow(table.Row{"col1"})
+    rows = append(rows, table.Row{"col1"})
     for _, node := range renderBatch {
         switch node.(type) {
         case float64, float32:
@@ -375,12 +380,14 @@ func renderSingleColumnTable(renderBatch []renderNode, opts *Opts) string {
             rows = append(rows, []interface{}{fmt.Sprintf("%v", node)})
         }
     }
-    t.AppendRows(rows)
-    var rowsAmount = t.Length()
-    t.Style().Options.SeparateRows = true
+
+    var rowsAmount = len(rows)
     if opts.TransposeMode {
-        t.TransposeAppendedRows()
+        rows = transposeRows(rows)
     }
+    t.AppendRows(rows)
+
+    t.Style().Options.SeparateRows = true
     if opts.NoGraphics {
         t.SetStyle(table.Style{Box: StyleWithoutGraphics})
     }
@@ -452,12 +459,41 @@ func castRenderNodeToMap(node renderNode) (map[string]renderNode, error) {
     return castedRes, nil
 }
 
+func transposeRows(rowsRaw []table.Row) []table.Row {
+    var rowsRawTransposedCap = 0
+    for _, row := range rowsRaw {
+        if len(row) > rowsRawTransposedCap {
+            rowsRawTransposedCap = len(row)
+        }
+    }
+
+    for rowNum, row := range rowsRaw {
+        if len(row) < rowsRawTransposedCap {
+            for i := 0; i < rowsRawTransposedCap-len(row); i++ {
+                rowsRaw[rowNum] = append(rowsRaw[rowNum], "")
+            }
+        }
+    }
+
+    var rowsRawTransposed []table.Row
+
+    for i := 0; i < rowsRawTransposedCap; i++ {
+        var rowTransposed table.Row
+        for j := 0; j < len(rowsRaw); j++ {
+            rowTransposed = append(rowTransposed, rowsRaw[j][i])
+        }
+        rowsRawTransposed = append(rowsRawTransposed, rowTransposed)
+    }
+
+    return rowsRawTransposed
+}
+
 // renderArrs returns table as string for arrays case.
 func renderArrs(renderBatch []renderNode, opts *Opts) (string, error) {
     t := table.NewWriter()
-    t.AppendRow(constructHeaderRowForArrs(renderBatch))
-
     var rows []table.Row
+
+    rows = append(rows, constructHeaderRowForArrs(renderBatch))
     for _, node := range renderBatch {
         nodeBytes, err := yaml.Marshal(node)
         if err != nil {
@@ -489,12 +525,14 @@ func renderArrs(renderBatch []renderNode, opts *Opts) (string, error) {
         }
         rows = append(rows, row)
     }
-    t.AppendRows(rows)
+
     t.Style().Options.SeparateRows = true
-    var rowsAmount = t.Length()
+    var rowsAmount = len(rows)
     if opts.TransposeMode {
-        t.TransposeAppendedRows()
+        rows = transposeRows(rows)
     }
+    t.AppendRows(rows)
+
     if opts.NoGraphics {
         t.SetStyle(table.Style{Box: StyleWithoutGraphics})
     }
```

Reply:
Done, thanks